File: Tracking/tracking.py
import numpy as np
from filterpy.kalman import KalmanFilter

def initialize_tracker():
    tracker = KalmanFilter(dim_x=4, dim_z=2)
    tracker.F = np.array([[1, 0, 1, 0],
                          [0, 1, 0, 1],
                          [0, 0, 1, 0],
                          [0, 0, 0, 1]])
    tracker.H = np.array([[1, 0, 0, 0],
                          [0, 1, 0, 0]])
    tracker.R = np.eye(2) * 0.1
    tracker.Q = np.eye(4) * 0.1
    return tracker

def update_tracks(tracker, measurement):
    tracker.predict()
    tracker.update(measurement)
    return tracker.x

File: Tracking/tracking.cpp
#include "tracking.hpp"

VectorXd initialize_tracker() {
    return VectorXd::Zero(4);
}

VectorXd update_tracks(MatrixXd& F, MatrixXd& H, MatrixXd& Q, MatrixXd& R, VectorXd& x, MatrixXd& P, const VectorXd& z) {
    // Predict
    x = F * x;
    P = F * P * F.transpose() + Q;

    // Update
    VectorXd y = z - H * x;
    MatrixXd S = H * P * H.transpose() + R;
    MatrixXd K = P * H.transpose() * S.inverse();
    x = x + K * y;
    P = (MatrixXd::Identity(4, 4) - K * H) * P;

    return x;
}
